scipy/special/_lambertw.h

```diff
@@ -0,0 +1,154 @@
+/* Implementation of the Lambert W function [1]. Based on MPMath
+ *  Implementation [2], and documentation [3].
+ *
+ * Copyright: Yosef Meller, 2009
+ * Authoer email: mellerf@netvision.net.il
+ *
+ * Distributed under the same license as SciPy
+ * Translated into C++ by SciPy developers, 2023.
+ *
+ * References:
+ * [1] On the Lambert W function, Adv. Comp. Math. 5 (1996) 329-359,
+ *     available online: http://www.apmaths.uwo.ca/~djeffrey/Offprints/W-adv-cm.pdf
+ * [2] mpmath source code, Subversion revision 990
+ *     http://code.google.com/p/mpmath/source/browse/trunk/mpmath/functions.py?spec=svn994&r=992
+ * [3] mpmath source code, Subversion revision 994
+ *     http://code.google.com/p/mpmath/source/browse/trunk/mpmath/function_docs.py?spec=svn994&r=994
+
+ * TODO: use a series expansion when extremely close to the branch point
+ * at `-1/e` and make sure that the proper branch is chosen there.
+ */
+
+
+#pragma once
+
+#include <cmath>
+#include <complex>
+#include <limits>
+
+#include "_evalpoly.h"
+#include "sf_error.h"
+
+using namespace std::complex_literals;
+
+constexpr double EXPN1 = 0.36787944117144232159553;  // exp(-1)
+constexpr double OMEGA = 0.56714329040978387299997;  // W(1, 0)
+
+
+inline std::complex<double> lambertw_branchpt(std::complex<double> z) {
+    // Series for W(z, 0) around the branch point; see 4.22 in [1].
+    double coeffs[] = {-1.0/3.0, 1.0, -1.0};
+    std::complex<double> p = std::sqrt(2.0*(M_E*z + 1.0));
+
+    return cevalpoly(coeffs, 2, p);
+}
+
+
+inline std::complex<double> lambertw_pade0(std::complex<double> z) {
+    // (3, 2) Pade approximation for W(z, 0) around 0.
+    double num[] = {
+	12.85106382978723404255,
+	12.34042553191489361902,
+	1.0
+    };
+    double denom[] = {
+	32.53191489361702127660,
+	14.34042553191489361702,
+	1.0
+    };
+
+    /* This only gets evaluated close to 0, so we don't need a more
+     * careful algorithm that avoids overflow in the numerator for
+     * large z. */
+    return z * cevalpoly(num, 2, z)/cevalpoly(denom, 2, z);
+}
+
+
+inline std::complex<double> lambertw_asy(std::complex<double> z, long k) {
+    /* Compute the W function using the first two terms of the
+     * asymptotic series. See 4.20 in [1].
+     */
+    std::complex<double> w;
+
+    w = std::log(z) + 2.0*M_PI*static_cast<double>(k)*1i;
```

Comment:
Do we need the `static_cast<double>(k)` here? Can we not use just `k`? We might, the complex version of `*` might complain without it, just checking.

Reply:
Thanks. I checked and it's not actually needed here. You can't directly multiply a `long` and a `std::complex<double>` but since the multiplication is evaluated left to right, that doesn't happen here. In an earlier version of the translation I was directly multiplying `k` with a complex value, got an error, and just blindly threw the `static_cast`s in everywhere to appease the compiler.

Reply:
Nice! Yes, I remember `std::complex<double>` plays badly with integers sometimes.